kore/src/Kore/Builtin/Map.hs

```diff
@@ -606,30 +622,139 @@ matchInKeys ::
     Maybe (InKeys (TermLike variable))
 matchInKeys = retract
 
-unifyNotInKeys ::
-    forall unifier.
-    MonadUnify unifier =>
-    TermSimplifier RewritingVariableName unifier ->
-    NotSimplifier unifier ->
+data UnifyNotInKeys = UnifyNotInKeys
+    { inKeys :: !(InKeys (TermLike RewritingVariableName))
+    , keyTerm, mapTerm :: !(TermLike RewritingVariableName)
+    , concreteKeys, mapKeys, opaqueElements :: ![TermLike RewritingVariableName]
+    , isFirstMatched :: !Bool
+    }
```

Comment:
I don't understand this datatype. It looks like `InKeys` has a field for the `keyTerm` and the `mapTerm`, but we also have separate `keyTerm` and `mapTerm` fields here. Also, what's with the `concreteKeys`, `mapKeys`, `opaqueElems`? Do these come from the `mapTerm`, if yes, then why separate them? Also, what's `isFirstMatched`?

Reply:
Those extra pieces of data are indeed derivable from the `normalizedMap`, but are needed to be inspected during the match in lines 667-674 and 693-701.  Since these pieces of data are also needed in the unification function, I figured it makes sense to package them in the record in the spirit of avoiding doing the same work twice.

Reply:
Also, `isFirstMatched` was removed, indeed that just some code I neglected to change earlier.